contracts/drafts/ERC777/ERC777.sol

```diff
@@ -336,22 +413,24 @@ contract ERC777 is IERC777 {
      * @param amount uint256 amount of tokens to transfer
      * @param userData bytes extra information provided by the token holder (if any)
      * @param operatorData bytes extra information provided by the operator (if any)
+     * @param avoidLockingTokens if true, contract recipients are required to implement ERC777TokensRecipient
      */
     function _callTokensReceived(
         address operator,
         address from,
         address to,
         uint256 amount,
         bytes memory userData,
-        bytes memory operatorData
+        bytes memory operatorData,
+        bool avoidLockingTokens
     )
     private
     {
         address implementer = _erc1820.getInterfaceImplementer(to, TOKENS_RECIPIENT_INTERFACE_HASH);
         if (implementer != address(0)) {
             IERC777Recipient(implementer).tokensReceived(operator, from, to, amount, userData, operatorData);
-        } else {
-            require(!to.isContract());
+        } else if (avoidLockingTokens) {
```

Comment:
According to my experiment, a small optimization that saves 10 gas for the branch that do not enter the IF block is:

if (allowLockingTokens) {
   return;
}
require(!to.isContract(), "ERC777: token recipient contract has no implementer for ERC777TokensRecipient");

This requires to change avoidLocking Tokens to its negated value allowLockingTokens. The non-ERC20 transactions will consume less gas, which looks logical to me since ERC-20 is only for compatibility. 10 gas seems small, but multiplied by the billions of transactions it will be a lot of gas spared.

Reply:
That's interesting @catageek!

10 gas is indeed very low though.